src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1225,17 +1239,27 @@ private void EnforceDoesNotReturn(SyntaxNode? syntaxOpt)
                 snapshotBuilder,
                 returnTypesOpt: null,
                 getFinalNullableState: false,
-                out _);
+                out _,
+                requiredAnalysis);
 
             var analyzedNullabilitiesMap = analyzedNullabilities.ToImmutable();
             var snapshotManager = snapshotBuilder?.ToManagerAndFree();
 
 #if DEBUG
             // https://github.com/dotnet/roslyn/issues/34993 Enable for all calls
-            if (compilation.IsNullableAnalysisEnabled)
+            if (snapshotManager != null && isNullableAnalysisEnabled(compilation))
```

Comment:
>snapshotManager != null [](start = 16, length = 23)

This feels wrong. We should be verifying the nullabilities map, regardless of whether we were taking snapshots at the time. This can be null during speculative analysis.